backend/src/appointment/utils.py

```diff
@@ -103,38 +106,76 @@ def chunk_list(to_chunk: list, chunk_by: int):
         yield to_chunk[i : i + chunk_by]
 
 
+def determine_database_driver(dialect: str) -> str:
+    """Return our preferred ORM driver for the given database dialect.
+
+    :param dialect: A class of database, like "mysql" or "postgresql".
+    :type dialect: str
+
+    :return: The string indicator of which driver the ORM should use.
+    :rtype: str
+    """
+
+    if dialect == 'mysql':
+        return 'mysqldb'
+    elif dialect == 'postgresql':
+        return 'psycopg2'
+    else:
+        return None
+
+
+@cache
+def get_database_url() -> str | sqlalchemy_url:
+    """Returns a database URL value suitable for passing to SQLAlechemy's ``create_engine`` function. If DATABASE_URL is
+    set in the environment, that string will be used, but this is not preferred as there are problems handling some
+    special characters. Otherwise (and preferably), uses SQLAlchemy's URL factory to produce working connection URL
+    objects in which special character escaping is handled for us.
+
+    :raises ValueError: If a necessary URL component is not configured.
+
+    :return: Either a string or a sqlalchemy.engine.url.URL, depending on configuration.
+    :rtype: str | sqlalchemy_url
+    """
+
+    # If a url is set directly, reluctantly pass that through
+    if 'DATABASE_URL' in os.environ:
+        log.info(
+            'The use of the DATABASE_URL environment variable is discouraged. Instead, use DATABASE_HOST, '
+            'DATABASE_PORT, DATABASE_USER, DATABASE_PASSWORD, DATABASE_ENGINE, and DATABASE_NAME.'
+        )
+        return os.environ.get('DATABASE_URL')
+
+    # But preferably pull the URL components from environment variables.
+    # These are settings we can safely assume defaults about.
+    db_name = os.environ.get('DATABASE_NAME', 'appointment')
+    dialect = os.environ.get('DATABASE_ENGINE', 'mysql')
+    driver = determine_database_driver(dialect=dialect)
+    port = int(os.environ.get('DATABASE_PORT', '3306'))
+
+    # These settings are secrets and must be set manually.
+    host = os.environ.get('DATABASE_HOST')
+    username = os.environ.get('DATABASE_USERNAME')
+    password = os.environ.get('DATABASE_PASSWORD')
+
+    if not db_name or not dialect or not host or not password or not port or not username:
```

Comment:
```suggestion
    if not all([db_name, dialect, host, password, port, username]):
```
Should be a slightly cleaner way of checking this.

Reply:
Yeah, I like that better. Will push this change.